extensions/scarb-execute/src/oracle/hint_processor.rs

```diff
@@ -0,0 +1,161 @@
+use cairo_vm::serde::deserialize_program::ApTracking;
+use cairo_vm::types::exec_scope::ExecutionScopes;
+use cairo_vm::vm::errors::hint_errors::HintError;
+use cairo_vm::vm::errors::vm_errors::VirtualMachineError;
+use cairo_vm::vm::runners::cairo_runner::{ResourceTracker, RunResources};
+use cairo_vm::vm::vm_core::VirtualMachine;
+use std::any::Any;
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+
+pub struct OracleHintProcessor<'a> {
+    pub cairo_hint_processor: CairoHintProcessor<'a>,
+    /// Whether `--experimental-oracles` flag has been enabled.
+    experiment_enabled: bool,
+}
+
+impl<'a> OracleHintProcessor<'a> {
+    /// Creates a new instance of [`OracleHintProcessor`].
+    pub fn new(cairo_hint_processor: CairoHintProcessor<'a>, experiment_enabled: bool) -> Self {
+        Self {
+            cairo_hint_processor,
+            experiment_enabled,
+        }
+    }
+
+    /// Gracefully look if this is one of the cheat codes supported by us.
+    /// Take over execution just right when we learn this is "our" cheat code.
+    /// Any errors will be created by the following [`CairoHintProcessor::execute_hint`] call.
+    fn execute_cheatcode(
+        &mut self,
+        vm: &mut VirtualMachine,
+        _exec_scopes: &mut ExecutionScopes,
+        hint_data: &Box<dyn Any>,
+        _constants: &HashMap<String, Felt252>,
+    ) -> ControlFlow<Result<(), HintError>> {
+        macro_rules! t {
+            ($expr:expr) => {
+                match $expr {
+                    Ok(v) => v,
+                    Err(e) => return ControlFlow::Break(Err(e.into())),
+                }
+            };
+        }
+
+        if let Some(Hint::Starknet(StarknetHint::Cheatcode {
+            selector,
+            input_start,
+            input_end,
+            output_start,
+            output_end,
+        })) = hint_data.downcast_ref::<Hint>()
+            && let Ok(selector) = str::from_utf8(&selector.value.to_bytes_be().1)
+            && selector == "oracle_invoke"
+        {
+            if !self.experiment_enabled {
+                return ControlFlow::Break(Err(HintError::AssertionFailed(
+                    "Oracles are experimental feature. \
+                    To enable, pass --experimental-oracles CLI flag."
+                        .into(),
+                )));
+            }
+
+            // Extract the inputs.
+            let input_start = t!(extract_relocatable(vm, input_start));
+            let input_end = t!(extract_relocatable(vm, input_end));
+            let inputs = t!(vm_get_range(vm, input_start, input_end));
+
+            // Prepare output segment.
+            let mut res_segment = MemBuffer::new_segment(vm);
+            let res_segment_start = res_segment.ptr;
+
+            // Route selector to particular execution methods.
+            match selector {
+                "oracle_invoke" => t!(self.execute_invoke(inputs, &mut res_segment)),
+                _ => return ControlFlow::Continue(()),
```

Comment:
This branch should be unreachable, as we have `&& selector == "oracle_invoke"` in `if`, right? 🤔